provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,211 @@
+	}
+	return slices.Contains(topLevelStableOutputs, propName)
+}
+
+func isReadOnly(readOnly []string, key string) bool {
+	for _, prop := range readOnly {
+		if prop == key {
+			return true
+		}
+	}
+	return false
+}
+
+// filterAndReturnNested will filter the readOnly nested properties and
+// return the nested properties if found
+// e.g.
+//   - if the readOnly properties are ["foo/bar", "foo/bar/baz", "somethingElse"]
+//     and the key is "foo"
+//     then the result will be ["bar", "bar/baz"]
+func filterAndReturnNested(readOnly []string, key string) []string {
+	result := []string{}
+	for _, prop := range readOnly {
+		if strings.HasPrefix(prop, key+"/*/") {
+			result = append(result, strings.TrimPrefix(prop, key+"/*/"))
+		} else if strings.HasPrefix(prop, key+"/") {
+			result = append(result, strings.TrimPrefix(prop, key+"/"))
+		}
+	}
+	return result
+}
+
+// previewOutputValue exists to recurse through nested objects and populate computed values
+// There are cases where the input and output types are the same, but some properties of the type
+// are only output values.
+func previewOutputValue(
+	inputValue resource.PropertyValue,
+	types map[string]metadata.CloudAPIType,
+	prop *schema.TypeSpec,
+	readOnly []string,
+) resource.PropertyValue {
+	switch {
+	case inputValue.IsSecret():
+		return resource.NewSecretProperty(&resource.Secret{
+			Element: previewOutputValue(inputValue.SecretValue().Element, types, prop, readOnly),
+		})
+	case inputValue.IsOutput():
+		return resource.NewOutputProperty(resource.Output{
+			Element:      previewOutputValue(inputValue.OutputValue().Element, types, prop, readOnly),
+			Known:        inputValue.OutputValue().Known,
+			Secret:       inputValue.OutputValue().Secret,
+			Dependencies: inputValue.OutputValue().Dependencies,
+		})
+	case inputValue.IsArray() && (prop.Type == "array" || prop.Items != nil):
+		var items []resource.PropertyValue
+		for _, item := range inputValue.ArrayValue() {
+			items = append(items, previewOutputValue(item, types, prop.Items, readOnly))
+		}
+		return resource.NewArrayProperty(items)
+	case inputValue.IsObject() && strings.HasPrefix(prop.Ref, "#/types/"):
+		typeName := strings.TrimPrefix(prop.Ref, "#/types/")
+		if t, ok := types[typeName]; ok {
+			v := previewResourceOutputs(inputValue.ObjectValue(), types, t.Properties, readOnly)
+			return resource.NewObjectProperty(v)
+		}
+	case inputValue.IsObject() && prop.AdditionalProperties != nil && (prop.AdditionalProperties.Ref != "" || prop.AdditionalProperties.Items != nil):
+		return previewOutputValue(inputValue, types, prop.AdditionalProperties, readOnly)
+	case inputValue.IsObject() && prop.AdditionalProperties != nil && prop.AdditionalProperties.Ref == "":
+		inputObject := inputValue.ObjectValue()
+		result := resource.PropertyMap{}
+		for name, value := range inputObject {
+			p := value
+			result[name] = previewOutputValue(p, types, prop.AdditionalProperties, readOnly)
```

Comment:
I suspect `readOnly` here is a bug, it has to be something like `mapElementReadOnly(readOnly)`, no? How does CF schema specify readOnly properties that are nested under maps? Is this a degenerate case that just does not occur?